### PR TITLE
Optimize empty IN subquery and empty INNER/RIGHT JOIN.

### PR DESCRIPTION
**Pull request by nicelulu**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):

- Performance Improvement

Short description (up to few sentences):

If the subquery is empty can quickly return.
https://github.com/yandex/ClickHouse/issues/6615

### CODE REVIEW

File: dbms/tests/queries/0_stateless/01012_empty_in_inner_right_join.sql [hunk 1]
```diff
@@ -0,0 +1,2 @@
+SELECT count() FROM system.numbers WHERE number IN (SELECT toUInt64(1) WHERE 0);
```
Comment (alexey-milovidov) [inline dbms/tests/queries/0_stateless/01012_empty_in_inner_right_join.sql]: Please also add a test case with RIGHT JOIN and LEFT JOIN.
And a case with NOT IN empty set.
And a case where IN set is not in WHERE clause: `SELECT avg(x IN (empty set)) FROM ...`
And a case when you have multiple sets, some of them are in WHERE and some of them are not.

File: dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp [hunk 2]
```diff
@@ -183,6 +188,13 @@ void CreatingSetsBlockInputStream::createOne(SubqueryForSet & subquery)
     }
     else
     {
+        if (subquery.set)
+            intersecting_with_empty &= true;
```
Comment (alexey-milovidov) [inline dbms/src/DataStreams/CreatingSetsBlockInputStream.cpp]: That looks weird.

File: dbms/tests/queries/0_stateless/01012_empty_in_inner_right_join.sql (hunk 1 above)
Comment (alexey-milovidov) [reply dbms/tests/queries/0_stateless/01012_empty_in_inner_right_join.sql]: And add a case like `WHERE NOT ignore(x IN empty set)`

File: dbms/src/Interpreters/Join.h [hunk 3]
```diff
@@ -175,6 +175,7 @@ class Join
     ASTTableJoin::Strictness getStrictness() const { return strictness; }
     AsofRowRefs::Type getAsofType() const { return *asof_type; }
     bool anyTakeLastRow() const { return any_take_last_row; }
+    bool innerOrRightKind() const { return kind == ASTTableJoin::Kind::Inner || kind == ASTTableJoin::Kind::Right; }
```
Comment (4ertus2) [inline dbms/src/Interpreters/Join.h]: ```isInner(kind) && isRight(kind)``` looks clearer.

File: dbms/src/DataStreams/CreatingSetsBlockInputStream.h [hunk 4]
```diff
@@ -43,11 +43,17 @@ class CreatingSetsBlockInputStream : public IBlockInputStream
     size_t rows_to_transfer = 0;
     size_t bytes_to_transfer = 0;
 
+    /// It's IN or INNER/RIGHT JOIN and set is empty.
```
Comment (alexey-milovidov) [inline dbms/src/DataStreams/CreatingSetsBlockInputStream.h]: Now the comment is obsolete.

Comment (nicelulu) [reply dbms/src/DataStreams/CreatingSetsBlockInputStream.h]: Done.

**Comment by alexey-milovidov** [comment]
Some tests have failed.

**Comment by nicelulu** [comment]
This is the normal result:
```
SELECT count()
FROM
(
    SELECT *
    FROM system.numbers
    LIMIT 1000
)
WHERE 1 IN
(
    SELECT 0
    WHERE 0
)

┌─count()─┐
│       0 │
└─────────┘

1 rows in set. Elapsed: 0.008 sec. Processed 1.00 thousand rows, 8.00 KB (117.63 thousand rows/s., 941.06 KB/s.)
```
This is the result of this pr:
```
SELECT count()
FROM
(
    SELECT *
    FROM system.numbers
    LIMIT 1000
)
WHERE 1 IN
(
    SELECT 0
    WHERE 0
)

Ok.

0 rows in set. Elapsed: 0.014 sec.
```
```
Expression
 Expression
  CreatingSets
   Lazy
   Aggregating
    Concat
     Expression
      Filter
       Limit
        Expression
         Expression
          Numbers
```
Maybe should not be doing ```CreatingSetsBlockInputStream``` return quickly? Because "conut 0" is ```Aggregating``` the return, should we do something  during IN or JOIN ```Expression```?

**Comment by nicelulu** [comment]
This is the normal result:
```
SELECT count()
FROM
(
    SELECT *
    FROM system.numbers
    LIMIT 1000
)
WHERE 1 IN
(
    SELECT 0
    WHERE 0
)

┌─count()─┐
│       0 │
└─────────┘

1 rows in set. Elapsed: 0.008 sec. Processed 1.00 thousand rows, 8.00 KB (117.63 thousand rows/s., 941.06 KB/s.)
```
This is the result of this pr:
```
SELECT count()
FROM
(
    SELECT *
    FROM system.numbers
    LIMIT 1000
)
WHERE 1 IN
(
    SELECT 0
    WHERE 0
)

Ok.

0 rows in set. Elapsed: 0.014 sec.
```
```
Expression
 Expression
  CreatingSets
   Lazy
   Aggregating
    Concat
     Expression
      Filter
       Limit
        Expression
         Expression
          Numbers
```
Maybe should not be doing CreatingSetsBlockInputStream return quickly? Because "conut 0" is Aggregating the return, should we do something before the IN or JOIN Expression?

**Comment by alexey-milovidov** [comment]
@nicelulu That becomes more interesting.
To get proper result for aggregation query, you have to provide empty data stream before Aggregating and not after CreatingSets. Let's think how to do it...

**Comment by alexey-milovidov** [comment]
> Maybe should not be doing CreatingSetsBlockInputStream return quickly?

Yes, this is not the correct way of implementation.
We need to check if the set is empty deeper in query execution pipeline.

**Comment by nicelulu** [comment]
@alexey-milovidov Maybe I can use the action sequence in ```ExpressionActions``` to determine an ```IN``` or ```JOIN```?

**Comment by nicelulu** [comment]
```
SELECT count()
FROM system.numbers 
WHERE (number IN 
(
    SELECT toUInt64(1)
    WHERE 0
)) AND (number != 1)
```

In fact, for ```IN```, I need to consider short-circuit ```AND``` operation. If it is ```OR```, I can't use this optimization.

**Comment by nicelulu** [comment]
For ```Functional stateless tests``` "00017_in_subquery_with_empty_result"
```
This PR result :
SET output_format_write_statistics = 0

Ok.

0 rows in set. Elapsed: 0.004 sec.
```
```
SELECT count()
FROM
(
    SELECT *
    FROM system.numbers
    LIMIT 1000
)
WHERE 1 IN
(
    SELECT 0
    WHERE 0
)
FORMAT JSON

{
	"meta":
	[
		{
			"name": "count()",
			"type": "UInt64"
		}
	],

	"data":
	[
		{
			"count()": "0"
		}
	],

	"rows": 1,

	"rows_before_limit_at_least": 0
}

1 rows in set. Elapsed: 0.015 sec.
```
00017_in_subquery_with_empty_result.reference result:
```
{
	"meta":
	[
		{
			"name": "count()",
			"type": "UInt64"
		}
	],

	"data":
	[
		{
			"count()": "0"
		}
	],

	"rows": 1,

	"rows_before_limit_at_least": 1000
}
```
Maybe it's normal?

**Comment by alexey-milovidov** [comment]
> Maybe it's normal?

Yes, it's perfect.

**Pull request merged**